### PR TITLE
contrib: add random connection balancer

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Loong Dai <loong.dai@intel.com>

Add random connection balancer, which is a replacement for exact connection balancer on Windows,

Fix TODO from @davinci26

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: contrib/network/connection_balance/random/source/connection_balancer_impl.h
Comment (KBaichoo) [inline contrib/network/connection_balance/random/source/connection_balancer_impl.h:18]: s/work threads, faster than Exact connection balancer. /worker threads. It is faster than the Exact connection balancer./

File: source/server/listener_impl.cc
Comment (KBaichoo) [inline source/server/listener_impl.cc:691]: prior for windows the code below never ran, will we run into issues now in the new version since it's not defined out?

**Review by KBaichoo — COMMENTED** [review]
Can we add some tests? Thanks!

File: contrib/network/connection_balance/random/source/connection_balancer_impl.cc
Comment (soulxu) [inline contrib/network/connection_balance/random/source/connection_balancer_impl.cc:29]: I don't think we need the lock here. and the handlers list won't be changed after the Envoy startup. Otherwise, I don't think it can provide better performance than the exact balancer.

**Review by soulxu — CHANGES_REQUESTED** [review]
A comment on the lock,

File: contrib/extensions_metadata.yaml
Comment (daixiang0) [reply contrib/extensions_metadata.yaml:81]: Done.

File: api/contrib/envoy/extensions/network/connection_balance/random/v3alpha/random.proto
Comment (daixiang0) [reply api/contrib/envoy/extensions/network/connection_balance/random/v3alpha/random.proto:13]: Added.

File: contrib/network/connection_balance/random/source/connection_balancer_impl.h
Comment (daixiang0) [reply contrib/network/connection_balance/random/source/connection_balancer_impl.h:18]: Updated.

File: source/server/listener_impl.cc
Comment (daixiang0) [reply source/server/listener_impl.cc:691]: The old logic:
- check WIN or not
- force use Exact connection balancer if WIN
- use one balancer refer to config if not WIN

The new logic:
- check config existed or not
- use one balancer refer to config if existed
- force use Exact connection balancer if WIN and config not existed

File: contrib/network/connection_balance/random/source/connection_balancer_impl.cc
Comment (daixiang0) [reply contrib/network/connection_balance/random/source/connection_balancer_impl.cc:29]: Delete it.

File: docs/root/configuration/other_features/random.rst
Comment (KBaichoo) [inline docs/root/configuration/other_features/random.rst:18]: forgot to update this

File: source/server/listener_impl.cc
Comment (KBaichoo) [reply source/server/listener_impl.cc:691]: so we now can incorrectly configure a connection balancer that doesn't work on WIN now, compared to before?

**Review by KBaichoo — COMMENTED** [review]
Thanks for working on this!

File: CODEOWNERS
Comment (daixiang0) [reply CODEOWNERS:288]: I am an end user from Intel.

File: docs/root/configuration/other_features/random.rst
Comment (daixiang0) [reply docs/root/configuration/other_features/random.rst:18]: My mistake, thanks.

**Comment by daixiang0** [comment]
> https://github.com/envoyproxy/envoy/pull/21957#discussion_r926936674

After this, we have two extend balancer: dlb and random. DLB only supports Linux x86, it will panic on WIN, so the case will not happen.

File: api/contrib/envoy/extensions/network/connection_balance/random/v3alpha/random.proto
Comment (daixiang0) [reply api/contrib/envoy/extensions/network/connection_balance/random/v3alpha/random.proto:17]: I update it as `Random connection balancer balances connections randomly among worker threads under all architectures.`

File: contrib/network/connection_balance/random/source/config.h
Comment (daixiang0) [reply contrib/network/connection_balance/random/source/config.h:8]: Changed.

File: CODEOWNERS
Comment (mattklein123) [reply CODEOWNERS:288]: I suspect you are not an end user as defined in that document. If you would like to discuss more privately to clarify feel free to email envoy-maintainers and we can discuss. Until then let's hold this pr. Thanks!

Comment (daixiang0) [reply CODEOWNERS:288]: Thank you, I will clarify in the mail.

**Pull request closed without merging**
